src/software/unix_full_system_main.cpp

```diff
@@ -14,12 +14,37 @@
 #include "software/constants.h"
 #include "software/estop/arduino_util.h"
 #include "software/logger/logger.h"
+#include "software/logger/proto_logger.h"
 #include "software/multithreading/observer_subject_adapter.hpp"
 #include "software/networking/udp/threaded_proto_udp_listener.hpp"
 #include "software/networking/unix/threaded_proto_unix_listener.hpp"
 #include "software/sensor_fusion/threaded_sensor_fusion.h"
 #include "software/util/generic_factory/generic_factory.h"
 
+// ProtoLogger has to be defined as a global variable so that it can be accessed by the
+// cleanup callback.
+std::shared_ptr<ProtoLogger> proto_logger;
+
+/**
+ * Signal handler which attempts to cleanly shutdown the program.
+ *
+ * @note This function is not guaranteed to clean up all resources since it calls
+ * non signal-safe functions (e.g. std::shared_ptr, std::cout, exit, etc.). The primary
+ * purpose of this function is to *try* to flush and stop the ProtoLogger.
+ *
+ * @param signal_num The signal number that was caught
+ */
+void cleanup(int signal_num)
```

Comment:
Took a look at how `g3log` handles flushing on fatal signals and it looks like they do they same thing -- ignore signal safety and pray nothing explodes

https://github.com/KjellKod/g3log/blob/master/src/crashhandler_unix.cpp

Reply:
That's good to know. G3log also has some utility function for signal handling and cleanup, but I ended up not using it initially since I was planning on flushing only after some signals, but then I changed my mind and kept the same `std::signal` implementation.